### PR TITLE
Support resuming of deepspeed + Lora + offloading

### PR DESCRIPTION
**Pull request by thepowerfuldeez**
This PR is a upstream version of @kazemf78 PR to support resuming of Lora training when using deepspeed.
Without setting `load_module_strict=False` as a default, checkpoint is not loaded due to Lora not containing all weights, throwing an error `deepspeed resume Error(s) in loading state_dict for PeftModelForCausalLM`

Related discussion: https://github.com/huggingface/peft/issues/746

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @pacman100 @younesbelkada

File: src/transformers/integrations/deepspeed.py [hunk 1]
```diff
@@ -414,7 +414,7 @@ def deepspeed_init(trainer, num_training_steps, inference=False):
     return optimizer, lr_scheduler
 
 
-def deepspeed_load_checkpoint(deepspeed_engine, checkpoint_path, load_module_strict=True):
+def deepspeed_load_checkpoint(deepspeed_engine, checkpoint_path, load_module_strict=False):
```
Comment (younesbelkada) [inline src/transformers/integrations/deepspeed.py:417]: can we somehow revert this and just force-set it to `True` in our trainer?

**Review by younesbelkada — COMMENTED** [review]
Thanks, in principle I would say this looks good, might be not ideal to change the default value of `load_module_strict` for the public method as other users might be using it. 
On DS side, I'll let @pacman100 comment on the PR 🙏

File: src/transformers/integrations/deepspeed.py (hunk 1 above)
Comment (thepowerfuldeez) [reply src/transformers/integrations/deepspeed.py:417]: That's a fair point! I'll push a change now

**Comment by kazemf78** [comment]
Could you please provide any updates on this PR?

**Comment by younesbelkada** [comment]
Sure @thepowerfuldeez ! 
@pacman100 is currently working on fixing issues with repsect to deepspeed and providing working scripts that you can run out of the box: https://github.com/huggingface/peft/pull/1489 we'll review this PR asap with sourab!

**Comment by pacman100** [comment]
Hello, this has been already fixed in https://github.com/huggingface/transformers/pull/28746. I ran experiments today and can confirm resuming training when using PEFT+DeepSpeed works

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1734,6 +1734,15 @@ def _inner_training_loop(
                 )
             elif is_sagemaker_mp_enabled() or self.is_fsdp_enabled:
                 self._load_from_checkpoint(resume_from_checkpoint, self.model_wrapped)
+        
+        # deepspeed ckpt loading
+        if resume_from_checkpoint is not None and self.is_deepspeed_enabled:
+            deepspeed_load_checkpoint(self.model_wrapped, resume_from_checkpoint, load_module_strict=False)
```
Comment (pacman100) [inline src/transformers/trainer.py:1740]: This is already happenging a couple lines above in 1732.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1734,6 +1734,15 @@ def _inner_training_loop(
...
+            if self.args.deepspeed_force_lr_scheduler_checkpointing and self.model_wrapped.lr_scheduler is None:
+                if os.path.isfile(os.path.join(resume_from_checkpoint, SCHEDULER_NAME)):
```
Comment (pacman100) [inline src/transformers/trainer.py:1742]: loading scheduler is handled in `_load_optimizer_and_scheduler` which is couple lines below

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -1316,6 +1316,18 @@ class TrainingArguments:
             "help": "Activates neftune noise embeddings into the model. NEFTune has been proven to drastically improve model performances for instrcution fine-tuning. Check out the original paper here: https://arxiv.org/abs/2310.05914 and the original code here: https://github.com/neelsjain/NEFTune. Only supported for `PreTrainedModel` and `PeftModel` classes."
         },
     )
+    
+    deepspeed_force_lr_scheduler_checkpointing: bool = field(
```
Comment (pacman100) [inline src/transformers/training_args.py:1320]: this isn;t required. Trainer saves the scheduler when it isn't part of DeepSpeed Engine. 
Below is a screenshot of a checkpoint saved with the sceduler file. So, this isn't required.

Screenshot 2024-02-22 at 4 17 27 PM

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -2416,6 +2425,12 @@ def _save_checkpoint(self, model, trial, metrics=None):
         else:
             staging_output_dir = os.path.join(run_dir, f"tmp-{checkpoint_folder}")
         self.save_model(staging_output_dir, _internal_call=True)
+        if self.is_deepspeed_enabled:
+            # under zero3 model file itself doesn't get saved since it's bogus! Unless deepspeed
+            # config `stage3_gather_16bit_weights_on_model_save` is True
+            self.model_wrapped.save_checkpoint(staging_output_dir)
```
Comment (pacman100) [inline src/transformers/trainer.py:2431]: this happens in `self.save_model(staging_output_dir, _internal_call=True)`

**Comment by ambroser53** [comment]
I'd like to bump this I'm running into the this issue on my project and it's causing significant delays. Does this branch solve the issue but is not tested enough for merging to main or is a solution still to be found?

**Comment by pacman100** [comment]
> I'd like to bump this I'm running into the this issue on my project and it's causing significant delays. Does this branch solve the issue but is not tested enough for merging to main or is a solution still to be found?

Please see my comments above, the transformers PR https://github.com/huggingface/transformers/pull/28746 should already be fixing this.
